src/test/java/com/yelp/nrtsearch/server/grpc/PrimaryRestartTests.java

```diff
@@ -110,7 +120,377 @@ public void testPrimaryRestartReplication() throws IOException {
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion2, "test_index", 1, 2, 3, 4, 5, 6, 7, 8);
+
+    primaryServer.addSimpleDocs("test_index", 12);
+    primaryServer.refresh("test_index");
+
+    replicaServer.waitForReplication("test_index");
+
+    primaryServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11, 12);
+    replicaServer.verifySimpleDocIds("test_index", 1, 2, 3, 9, 10, 11, 12);
+
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion1, "test_index", 1, 2, 3, 4, 5);
+    verifySimpleDocIdsForVersion(
+        replicaServer, previousSearcherVersion2, "test_index", 1, 2, 3, 4, 5, 6, 7, 8);
+  }
+
+  @Test
+  public void testCleanupPreviousSearcher() throws IOException, InterruptedException {
+    TestServer primaryServer =
+        TestServer.builder(folder)
+            .withAutoStartConfig(true, Mode.PRIMARY, 0, IndexDataLocationType.REMOTE)
+            .withWriteDiscoveryFile(true)
+            .build();
+    primaryServer.createSimpleIndex("test_index");
+    primaryServer.startPrimaryIndex("test_index", -1, null);
+
+    primaryServer.addSimpleDocs("test_index", 1, 2, 3);
+    primaryServer.commit("test_index");
+    primaryServer.refresh("test_index");
+
+    TestServer replicaServer =
+        TestServer.builder(folder)
+            .withAutoStartConfig(true, Mode.REPLICA, -1, IndexDataLocationType.REMOTE)
+            .withAdditionalConfig(
+                String.join(
+                    "\n",
+                    "discoveryFileUpdateIntervalMs: 1000",
+                    "filterIncompatibleSegmentReaders: true"))
+            .build();
+
+    primaryServer.addSimpleDocs("test_index", 4, 5);
+    primaryServer.refresh("test_index");
+
+    replicaServer.waitForReplication("test_index");
+    replicaServer.verifySimpleDocIds("test_index", 1, 2, 3, 4, 5);
+    long previousSearcherVersion1 =
+        replicaServer
+            .getClient()
+            .getBlockingStub()
+            .search(
+                SearchRequest.newBuilder()
+                    .setIndexName("test_index")
+                    .setQuery(Query.newBuilder().build())
+                    .build())
+            .getSearchState()
+            .getSearcherVersion();
+    IndexSearcher previousSearcher1 =
+        replicaServer
+            .getGlobalState()
+            .getIndex("test_index")
+            .getShard(0)
+            .slm
+            .acquire(previousSearcherVersion1);
+    List<LeafReaderContext> previousLeaves1 = previousSearcher1.getIndexReader().leaves();
+    replicaServer
+        .getGlobalState()
+        .getIndex("test_index")
+        .getShard(0)
+        .slm
+        .release(previousSearcher1);
+
+    primaryServer.addSimpleDocs("test_index", 6, 7, 8);
+    primaryServer.refresh("test_index");
+
+    replicaServer.waitForReplication("test_index");
+    replicaServer.verifySimpleDocIds("test_index", 1, 2, 3, 4, 5, 6, 7, 8);
+    long previousSearcherVersion2 =
```

Comment:
The searcher versions and searchers are retrieved and searcher released several times in this test. Can extract them into methods to make this more readable.